File: groupproperties.rb
require 'gtk2'
require 'components/group'

module Cpg
	class GroupProperties < Gtk::Table
		type_register
		
		attr_reader :combo_main, :combo_class
	
		def initialize(objs, defmain = nil, defklass = nil)
			super({
					'n_rows' => 2, 
					'n_columns' => 2, 
					'homogeneous' => false, 
					'row_spacing' => 3, 
					'column_spacing' => 3
			})
		
			@objs = objs.select { |x| not x.is_a?(Components::Attachment) }
			build(defmain, defklass)
			show_all
		end
	
		def main
			@combo_main.active_iter ? @combo_main.active_iter[0] : nil
		end
	
		def klass
			@combo_class.active_text
		end
		
		def make_label(lbl)
			l = Gtk::Label.new(lbl)
			l.xalign = 0
			
			l
		end
	
		def build(defmain, defklass)
			attach(make_label('Relay:'), 0, 1, 0, 1, Gtk::FILL, Gtk::FILL)
			attach(make_label('Class:'), 0, 1, 1, 2, Gtk::FILL, Gtk::FILL)
		
			store = Gtk::ListStore.new(Object)
			activeidx = 0
			idx = 1
			
			piter = store.append
			piter[0] = nil
			
			@objs.each do |obj|
				next unless obj.property_set?('id')

				piter = store.append
				piter[0] = obj
				
				if obj == defmain
					activeidx = idx
				end
				
				idx = idx + 1
			end
		
			combo = Gtk::ComboBox.new(store)
			renderer = Gtk::CellRendererText.new
			combo.pack_start(renderer, true)
			combo.set_cell_data_func(renderer) do |layout, cell, model, piter|
				cell.text = piter[0] ? piter[0].to_s : 'None'
			end
			combo.active = activeidx
			@combo_main = combo
		
			attach(combo, 1, 2, 0, 1)
			
			activeidx = 0
			idx = 0
			
			combo = Gtk::ComboBox.new
			Components::Groups::Renderer.classes.each do |klass|
				cname = klass.to_s.gsub(/.*::/, '')
				combo.append_text(cname)
				
				if cname == defklass.to_s
					activeidx = idx
				end
				
				idx = idx + 1
			end
			
			combo.active = activeidx
			@combo_class = combo
		
			attach(combo, 1, 2, 1, 2)
		end
	end
end

File: GroupProperties.cs
using System;
using Gtk;
using System.Collections.Generic;
using System.Reflection;

namespace Cpg.Studio
{
	public class GroupProperties : Gtk.Table
	{
		private List<Components.Simulated> d_objects;
		private ComboBox d_comboMain;
		private ComboBox d_comboKlass;
		
		public GroupProperties(Components.Object[] objects, Components.Simulated defmain, Type defklass) : base(2, 2, false)
		{
			RowSpacing = 6;
			ColumnSpacing = 12;
			
			/* Get all simulated non-links */
			d_objects = new List<Components.Simulated>();
			
			foreach (Components.Object obj in objects)
			{
				if (obj is Components.Simulated && !(obj is Components.Link))
					d_objects.Add(obj as Components.Simulated);
			}
			
			Build(defmain, defklass);
			ShowAll();
		}
		
		public GroupProperties(Components.Object[] objects) : this(objects, null, null)
		{
		}
		
		private Widget MakeLabel(string text)
		{
			Label label = new Label(text);
			label.Xalign = 0;
			
			return label;
		}
		
		private Gdk.Pixbuf GroupIcon(Type type)
		{
			ConstructorInfo info = type.GetConstructor(new Type[] {});
			
			if (info == null)
				return null;
			
			Components.Renderers.Group renderer = info.Invoke(new object[] {}) as Components.Renderers.Group;
			
			return renderer.Icon(24);
		}
		
		private void Build(Components.Simulated defmain, Type defklass)
		{
			Attach(MakeLabel("Relay:"), 0, 1, 0, 1, AttachOptions.Fill, AttachOptions.Fill, 0, 0);
			Attach(MakeLabel("Class:"), 0, 1, 1, 2, AttachOptions.Fill, AttachOptions.Fill, 0, 0);
			
			ListStore store = new ListStore(typeof(Components.Simulated), typeof(String), typeof(Gdk.Pixbuf));
			TreePath path = null;
			
			foreach (Components.Simulated obj in d_objects)
			{
				if (String.IsNullOrEmpty(obj.Id))
					continue;
				
				TreeIter iter = store.Append();
				store.SetValue(iter, 0, new GLib.Value(obj));
				store.SetValue(iter, 1, new GLib.Value(obj.Id));
				
				if (obj.Renderer != null)
					store.SetValue(iter, 2, new GLib.Value(obj.Renderer.Icon(24)));
				
				if (obj == defmain)
					path = store.GetPath(iter);
			}
			
			ComboBox cmb = new ComboBox(store);
			CellRendererPixbuf iconrenderer = new CellRendererPixbuf();
			cmb.PackStart(iconrenderer, false);
			cmb.AddAttribute(iconrenderer, "pixbuf", 2);
			
			CellRendererText textrenderer = new CellRendererText();
			cmb.PackStart(textrenderer, true);
			cmb.AddAttribute(textrenderer, "text", 1);
			textrenderer.Yalign = 0.5f;
			
			if (path != null)
			{
				TreeIter it;
				store.GetIter(out it, path);
				cmb.SetActiveIter(it);
			}
			else
			{
				cmb.Active = 0;
			}
			
			d_comboMain = cmb;
			Attach(cmb, 1, 2, 0, 1);
			
			store = new ListStore(typeof(Type), typeof(String), typeof(Gdk.Pixbuf));
			cmb = new ComboBox(store);
			path = null;
			
			Assembly asm = Assembly.GetEntryAssembly();
			
			foreach (Type type in asm.GetTypes())
			{
				if (!type.IsSubclassOf(typeof(Components.Renderers.Group)))
					continue;

				object[] attributes = type.GetCustomAttributes(typeof(Components.Renderers.NameAttribute), true);
				string name;
				
				if (attributes.Length != 0)
				{
					name = (attributes[0] as Components.Renderers.NameAttribute).Name;
				}
				else
				{
					name = "None";	
				}

				TreeIter iter = store.Append();
				
				store.SetValue(iter, 0, new GLib.Value(type));
				store.SetValue(iter, 1, new GLib.Value(name));
				store.SetValue(iter, 2, new GLib.Value(GroupIcon(type)));

				if (type == defklass)
					path = store.GetPath(iter);
			}
			
			iconrenderer = new CellRendererPixbuf();
			cmb.PackStart(iconrenderer, false);
			cmb.AddAttribute(iconrenderer, "pixbuf", 2);

			textrenderer = new CellRendererText();
			cmb.PackStart(textrenderer, true);
			cmb.AddAttribute(textrenderer, "text", 1);
			textrenderer.Yalign = 0.5f;
			
			if (path != null)
			{
				TreeIter it;
				store.GetIter(out it, path);
				cmb.SetActiveIter(it);
			}
			else
			{
				cmb.Active = 0;	
			}
			
			d_comboKlass = cmb;
			Attach(cmb, 1, 2, 1, 2);
		}
		
		public Components.Simulated Main
		{
			get
			{
				TreeIter iter;
				
				if (d_comboMain.GetActiveIter(out iter))
					return d_comboMain.Model.GetValue(iter, 0) as Components.Simulated;
				else
					return null;
			}
		}
		
		public Type Klass
		{
			get
			{
				TreeIter iter;
				
				if (d_comboKlass.GetActiveIter(out iter))
					return d_comboKlass.Model.GetValue(iter, 0) as Type;
				else
					return null;
			}
		}
		
		public ComboBox ComboMain
		{
			get
			{
				return d_comboMain;
			}
		}
		
		public ComboBox ComboKlass
		{
			get
			{
				return d_comboKlass;
			}
		}
	}
}
